File: ui/src/features/handoff/informants.ts
// Who told us what: the report's `informants` (herald/reporting/informants.py). A recording or a server from before
// 2026-09-26 does not send them, so the same grouping is derived here from the confirmed lines' sources: the people
// other than the crew first, then the patient monitor, then the crew, and what could not be attributed last.
import type { HandoffInformant, HandoffReportData, HandoffSection } from "@/lib/types";

// How a source is named when no speaker word is used: the server's config/handoff.yaml `informants.labels`
const WHO: Record<string, string> = {
  medic: "medic", patient: "patient", device: "patient monitor", photo: "photo", family: "family",
  bystander: "bystander", unknown: "speaker not identified",
};
// the reading order (herald/reporting/informants.py _ORDER)
const ORDER: Record<string, number> = { family: 0, bystander: 0, patient: 1, device: 2, photo: 3, medic: 4, unknown: 5 };

/** [{who, role, keys, items}] from the confirmed lines' sources; `label` names a key ("Medications"). */
export function deriveInformants(sections: HandoffSection[], label: (key: string) => string): HandoffInformant[] {
  const groups = new Map<string, HandoffInformant>();
  for (const line of sections.flatMap((section) => section.lines)) {
    if (line.status !== "confirmed") continue;
    for (const src of line.sources ?? []) {
      const role = src.role || "unknown";
      const named = (role === "family" || role === "bystander") && !!src.speaker;
      const id = named ? `${role}:${src.speaker}` : role;
      const group = groups.get(id) ?? { who: named ? String(src.speaker) : WHO[role] ?? role, role, keys: [], items: [] };
      groups.set(id, group);
      if (!group.keys.includes(src.key)) group.keys.push(src.key);
    }
  }
  const rows = [...groups.values()].sort((a, b) => (ORDER[a.role] ?? 0) - (ORDER[b.role] ?? 0));
  for (const row of rows) row.items = [...new Set(row.keys.map(label))];
  return rows;
}

/** The server's list when it sent one; otherwise the same list derived from the report's lines. */
export function reportInformants(r: HandoffReportData, label: (key: string) => string): HandoffInformant[] {
  return r.informants ?? deriveInformants(r.sections, label);
}

File: herald/reporting/informants.py
"""Who told us what: the handoff report's sources, grouped by who gave the information ("husband: medications,
allergies"; "patient monitor: heart rate, SpO2"; "medic: the rest").

Built only from the confirmed facts behind the report's own lines, so it never names a source for something the
report does not say. Who a fact came from is its provenance: the monitor for a camera reading of the patient monitor,
the medic, the patient, or the person the check step read it as from the words ("husband"); a room-mic fact whose
speaker the words did not show is listed as not identified. The wording is config/handoff.yaml `informants`.
"""
from __future__ import annotations

from ..core.vocabulary import Vocabulary

# the order a report reads its sources in: people other than the crew first (that is what the ED cannot see), then
# the devices, then the crew, and what could not be attributed last
_ORDER = {"family": 0, "bystander": 0, "patient": 1, "device": 2, "photo": 3, "medic": 4, "unknown": 5}


def _who(source: dict, words: dict) -> tuple[str, str]:
    """(group id, how the report names them) for one fact's source."""
    role = source.get("role") or "unknown"
    if role in ("family", "bystander") and source.get("speaker"):
        return f"{role}:{source['speaker']}", str(source["speaker"])
    return role, words["labels"].get(role, role)


def informants(sections: list[dict], vocab: Vocabulary, words: dict) -> list[dict]:
    """[{"who", "role", "keys", "items"}] in reading order; `items` are the vocabulary labels of what they told."""
    groups: dict[str, dict] = {}
    for section in sections:
        for line in section.get("lines", []):
            if line.get("status") != "confirmed":
                continue
            for src in line.get("sources", []):
                gid, who = _who(src, words)
                g = groups.setdefault(gid, {"who": who, "role": src.get("role") or "unknown", "keys": []})
                if src["key"] not in g["keys"]:
                    g["keys"].append(src["key"])
    out = sorted(groups.values(), key=lambda g: _ORDER.get(g["role"], 0))
    for g in out:
        items: list[str] = []
        for key in g["keys"]:
            label = vocab.label(key) if key in vocab else key
            if label not in items:
                items.append(label)
        g["items"] = items
    return out


def informants_sentence(rows: list[dict], words: dict, sep: str, list_sep: str) -> str:
    """"Who told us: husband: medications, allergies; patient monitor: heart rate; medic: the rest." When someone
    other than the crew told something, the crew's own long list is read as "the rest"."""
    if not rows:
        return ""
    others = any(r["role"] != "medic" for r in rows)
    parts = [f"{r['who']}: {words['rest'] if r['role'] == 'medic' and others else list_sep.join(r['items']).lower()}"
             for r in rows]
    return f"{words['heading']}: {sep.join(parts)}."
